tools/runner/logsaver.go

```diff
@@ -1,145 +1,103 @@
+// the log to the a given path if the container have logs. The name
+// of the files are in format pod-name-container-name.log. After
+// writing the logs the function returns a list of pointers of
+// LogInfo objects containing the log's information.
+func SaveAllLogs(ctx context.Context, loadTest *grpcv1.LoadTest, podsGetter corev1types.PodsGetter, pods []*corev1.Pod, podLogDir string) ([]*LogInfo, error) {
+	var logInfos []*LogInfo
 
 	// Attempt to create directory. Will not error if directory already exists
-	err = os.MkdirAll(podLogDir, os.ModePerm)
+	err := os.MkdirAll(podLogDir, os.ModePerm)
 	if err != nil {
-		return savedLogs, fmt.Errorf("Failed to create pod log output directory %s: %v", podLogDir, err)
+		return logInfos, fmt.Errorf("Failed to create pod log output directory %s: %v", podLogDir, err)
 	}
 
 	// Write logs to files
 	for _, pod := range pods {
-		logFilePath := filepath.Join(podLogDir, pod.Name+".log")
-		buffer, err := ls.getPodLogBuffer(ctx, pod)
-		if err != nil {
-			return savedLogs, fmt.Errorf("could not get log from pod: %s", err)
+		for _, container := range pod.Spec.Containers {
+			logFilePath := filepath.Join(podLogDir, fmt.Sprintf("%s-%s.log", pod.Name, container.Name))
+
+			logInfo, err := SaveLog(ctx, loadTest, pod, podsGetter, container.Name, logFilePath)
+			if err != nil {
+				return logInfos, fmt.Errorf("could not get log from container: %v", err)
+			}
+			if logInfo != nil {
+				logInfos = append(logInfos, logInfo)
+			}
 		}
-		err = ls.writeBufferToFile(buffer, logFilePath)
-		if err != nil {
-			return savedLogs, fmt.Errorf("could not write pod log buffer to file: %s", err)
-		}
-		savedLogs.podToPathMap[pod] = logFilePath
 	}
-	return savedLogs, nil
+	return logInfos, nil
 }
 
-// getTestPods retrieves the pods associated with a LoadTest.
-func (ls *LogSaver) getTestPods(ctx context.Context, loadTest *grpcv1.LoadTest) ([]*corev1.Pod, error) {
-	podLister := ls.podsGetter.Pods(metav1.NamespaceAll)
-
-	// Get a list of all pods
-	podList, err := podLister.List(ctx, metav1.ListOptions{})
-	if err != nil {
-		return nil, errors.New("Failed to fetch list of pods")
-	}
-
-	// Get pods just for this specific test
-	testPods := status.PodsForLoadTest(loadTest, podList.Items)
-	return testPods, nil
-}
-
-func (ls *LogSaver) getPodLogBuffer(ctx context.Context, pod *corev1.Pod) (*bytes.Buffer, error) {
-	req := ls.podsGetter.Pods(pod.Namespace).GetLogs(pod.Name, &corev1.PodLogOptions{})
-	driverLogs, err := req.Stream(ctx)
+// SaveLog retrieves and save logs for a specific container.
+//
+// SaveLog retrieves logs from a container under given container
+// name within given pod,then save the logs to the given file
+// path, if there are logs to save. The function also returns
+// a point of the LogInfo object.
+func SaveLog(ctx context.Context, loadTest *grpcv1.LoadTest, pod *corev1.Pod, podsGetter corev1types.PodsGetter, containerName string, filePath string) (*LogInfo, error) {
```

Comment:
If you made this function take the directory rather than the filepath, then you would be able to construct the filepath inside this function.

In fact, I would construct the logfilename in a separate function, from the pod name and the container name. Then inside SaveLog I would join that with the directory to construct the LogInfo.